main/pipeline/pipeline.cc

```diff
@@ -940,17 +945,28 @@ ast::ParsedFilesOrCancelled typecheck(unique_ptr<core::GlobalState> &gs, vector<
 
         {
             ProgressIndicator cfgInferProgress(opts.showProgress, "CFG+Inference", what.size());
-            workers.multiplexJob("typecheck", [ctx, &opts, fileq, resultq]() {
+            workers.multiplexJob("typecheck", [ctx, &opts, &typecheckMutex = gs->typecheckMutex, fileq, resultq,
+                                               cancelable, preemptible, epoch]() {
                 typecheck_thread_result threadResult;
                 ast::ParsedFile job;
                 int processedByThread = 0;
 
                 {
                     for (auto result = fileq->try_pop(job); !result.done(); result = fileq->try_pop(job)) {
                         if (result.gotItem()) {
+                            unique_ptr<absl::ReaderMutexLock> lock;
+                            if (preemptible) {
+                                // Acquire a reader lock here. Parks the thread if the typechecker thread is trying to
+                                // grab the lock to preempt typechecking.
+                                lock = make_unique<absl::ReaderMutexLock>(typecheckMutex.get());
```

Comment:
The fact that user needs to remember to correctly grab locks bothers me.
Can we make an abstraction where api forces users to use it right?
https://github.com/sorbet/sorbet/pull/2407/files#r362692618

Reply:
So worker thread Unfreeze's GlobalState before doing work which 'does the right thing' and grabs the reader lock?

Reply:
I can look into how the existing Unfreeze classes work.